File: lstore/table.cpp
#include <vector>
#include <map>
#include <string>
#include "RID.h"
#include "index.h"
#include "page.h"
#include "table.h"


Table::Table(std::string name, int key, int num_columns): name(name), key(key), num_columns(num_columns) {
    this->index = new Index();
    index->setTable(this);
};

/***
 *
 * Insert a record into appropriate base page.
 *
 * @param Record record A record to insert
 * @return RID of the new record upon successful insertion.
 *
 */
RID Table::insert(Record record) {
    if (pages.size() == 0 || pages.base_has_capacity()) {
        pages.pushback(PageRange(record)); // Make a base page with given record
        // return the RID for index or something
        return pages.back().page_range[0].first;
    } else { // If there are base page already, just insert it normally.
        return pages.back().insert(record);
    }
}

/***
 *
 * Given a RID to the original base page, column number, and new value, it will update by creating new entry on tail page.
 *
 * @param RID rid Rid that pointing to the base page.
 * @param int column Column to update
 * @param int new_value The value to update to
 * @return RID of the new row upon successful update
 *
 */
RID Table::update(RID rid, int column, int new_value) {
    int i = 0;
    for (; i < pages.size(); i++) {
        if (pages[i].page_range[0].first.id > rid.id) {
            break;
        }
    }
    i--;
    return pages[i].update(rid, column, new_value);
}

/***
 *
 * Merge few version of records and create new base page.
 *
 * Possible param : number of versions to merge
 *
 */
int Table::merge() {
    return -1;
}

File: lstore/table.py
from lstore.index import Index
from lstore.page import Page
from lstore.config import *
from time import time

# One record has a Page per column
class Record:

    def __init__(self, rid, key, columns):
        self.rid = rid
        self.key = key
        self.columns = columns
        self.indirection = -1
        self.schema_encoding = []

    def get_key(self):
        return self.columns[self.key]

class Table:

    """
    :param name: string         #Table name
    :param key: int             #Index of table key in columns
    :param num_columns: int     #Number of Columns: all columns are integer
    :param page_directory: dict #Page directory
    :param num_page_ranges: int #Number of page ranges
    """
    def __init__(self, name, num_columns, key):
        self.name = name
        self.key = key
        self.num_columns = num_columns
        self.page_directory = {}
        self.page_range = BASE_PAGE_MAX
        self.tail_pages = []
        self.farthest = {'pi': 0, 'slot_index': -1}
        for i in range(self.page_range):
            self.page_directory[i] = {}
            page_stack = self.page_directory[i]
            page_stack[0] = Page()
            page_stack[0].id = i
            page_stack['size'] = 1
        self.index = Index(self)

    def add_tail(self, pi):
        tail = Page()
        page_stack = self.page_directory[pi]
        tail.id = page_stack[0].id
        page_stack[page_stack['size']] = tail
        page_stack['size'] += 1
        return tail

    def __merge(self):
        print("merge is happening")
        pass
 
